silx/opencl/backprojection.py

```diff
@@ -0,0 +1,477 @@
+            #  cl.enqueue_copy(queue, dst, src, host_origin=(0,0), buffer_origin=(0,0), region=shape, host_pitches=(sino.shape[1],), buffer_pitches=(self.fft_size,))
+            # However it does not work properly, and raises an error for pyopencl < 2017.1
+            sino_zeropadded = np.zeros((sino.shape[0], self.fft_size), dtype=np.complex64)
+            sino_zeropadded[:, :self.num_bins] = sino.astype(np.float32)
+            sino_zeropadded = np.ascontiguousarray(sino_zeropadded, dtype=np.complex64)
+            # send to GPU
+            ev = pyopencl.enqueue_copy(self.queue, self.d_sino_z.data, sino_zeropadded)
+            #~ del sino_zeropadded
+            events.append(EventDescription("Send sino H->D", ev))
+            # FFT (in-place)
+            self.pyfft_plan.execute(self.d_sino_z.data, batch=self.num_projs)
+            # Multiply (complex-wise) with the the filter
+            ev = self.prg_mult.mult(self.queue, self.d_sino_z.shape[::-1], None,
+                           self.d_sino_z.data,
+                           self.d_filter.data,
+                           np.int32(self.fft_size),
+                           self.num_projs
+                         )
+            events.append(EventDescription("complex 2D-1D multiplication", ev))
+            # Inverse FFT (in-place)
+            self.pyfft_plan.execute(self.d_sino_z.data, batch=self.num_projs, inverse=True)
+            # Copy the real part of d_sino_z[:, :self.num_bins] (complex64) to d_sino (float32)
+            ev = self.prg_cpy2d.cpy2d(self.queue, self.shape[::-1], None,
+                           self.d_sino,
+                           self.d_sino_z.data,
+                           self.num_bins,
+                           self.num_projs,
+                           np.int32(self.fft_size)
+                         )
+            events.append(EventDescription("conversion from complex padded sinogram to sinogram", ev))
+            if self.profile:
+                self.events += events
+            # ------
+        else: # no pyfft
+            # Zero-padding of the sinogram
+            sino_zeropadded = np.zeros((sino.shape[0], self.fft_size), dtype=np.complex64)
+            sino_zeropadded[:, :self.num_bins] = sino.astype(np.float32)
+            # Linear convolution
+            sino_f = np.fft.fft(sino, self.fft_size)
+            sino_f = sino_f * self.filter
+            sino_filtered = np.fft.ifft(sino_f)[:, :self.num_bins].real
+            # Send the filtered sinogram to device
+            sino_filtered = np.ascontiguousarray(sino_filtered, dtype=np.float32)
+            pyopencl.enqueue_copy(self.queue, self.d_sino, sino_filtered)
+
+
+    def filtered_backprojection(self, sino):
+        """
+        Compute the filtered backprojection (FBP) on a sinogram.
+
+        :param sino: sinogram (`numpy.ndarray`) in the format (projections, bins)
+        """
+
+        self.filter_projections(sino)
```

Comment:
as filter_projection alters the device memory, it should be locked ...
If this method is the "main", the best would be to lock here (and not in backprojection). 
Unless 2 locks are advices to prevent dead-locks.